src/te/autodiff/ad_simplify.cc

```diff
@@ -0,0 +1,1305 @@
+  }
+
+  return arith::IntConstraintsTransform(domain, new_domain, src_to_dst, dst_to_src);
+}
+
+// Simplify an iteration domain.
+inline arith::IntConstraintsTransform IdentityTransformation(const arith::IntConstraints& domain) {
+  Map<Var, PrimExpr> identity_map;
+  for (const Var& v : domain->variables) {
+    identity_map.Set(v, v);
+  }
+  return arith::IntConstraintsTransform(domain, domain, identity_map, identity_map);
+}
+
+arith::IntConstraintsTransform SimplifyDomain(const arith::IntConstraints& iter_domains,
+                                              bool eliminate_div_mod) {
+  arith::IntConstraintsTransform transf = IdentityTransformation(iter_domains);
+
+  if (eliminate_div_mod) {
+    transf = transf + EliminateDivModFromDomainConditions(transf->dst);
+  }
+
+  // TODO(sgrechanik-h): Repeating the following steps has a positive effect, however we probably
+  // should find a better terminating criterion (like stop when the domain volume stops decreasing)
+  // Also 2 steps seems to be slightly better than 3
+  for (size_t i = 0; i < 2; ++i) {
+    arith::IntConstraintsTransform tr = arith::SolveLinearEquations(transf->dst);
+    transf = transf + tr;
+    // TODO(sgrechanik-h): This helps for some artificial examples, however I'm not sure about
+    // enabling it in general. The problem it solves is propagating equalities of outer vars.
+    // tr = AddOuterVariablesIntoDomain(transf->dst);
+    tr = arith::SolveInequalitiesDeskewRange(transf->dst);
+    transf = transf + tr;
+  }
+
+  return transf;
+}
+
+// Use the condition of a reduction op to simplify its domain (axis)
+PrimExpr SimplifyReductionDomain(const PrimExpr& expr, const Map<Var, Range>& outer_vranges) {
+  if (const ReduceNode* red = expr.as<ReduceNode>()) {
+    Array<Var> vars = IterVarsToVars(red->axis);
+    Map<Var, Range> vranges = Merge(outer_vranges, IterVarsToMap(red->axis));
+    Array<PrimExpr> relations = FactorOutAtomicFormulas(red->condition).to_array();
+
+    arith::IntConstraints domain(vars, vranges, relations);
+    auto res = SimplifyDomain(domain);
+
+    Array<PrimExpr> new_source;
+    for (const PrimExpr& src : red->source) {
+      new_source.push_back(Substitute(src, res->src_to_dst));
+    }
+
+    Array<IterVar> new_axis = IterVarsFromMap(res->dst->variables, res->dst->ranges, kCommReduce);
+
+    // Perform simplification mainly to remove a possibly empty reduction.
+    arith::Analyzer analyzer;
+    return analyzer.Simplify(
+        Reduce(red->combiner, new_source, new_axis, All(res->dst->relations), red->value_index),
+        ARITH_SIMPLIFY_REWRITE_CANONICAL_REWRITE);
+  } else {
+    return expr;
+  }
+}
+
+// Extract from cond an implication of cond not containing vars
+std::pair<PrimExpr, PrimExpr> ImplicationNotContainingVars(
+    const PrimExpr& cond, const std::unordered_set<const VarNode*>& vars) {
+  CHECK(cond.dtype().is_bool()) << "The type of cond must be bool";
+  // TODO(sgrechanik-h): NOT
```

Comment:
Fix TODO

Reply:
Actually in my understanding it's not straightforward to separate NOT node here, as the false branch of (!pair.a) will also contain the reduction (instead the zero). I'm not sure whether it provides benefit, @sergei-grechanik would you help to comment?

Reply:
I just have no idea why is there a single NOT floating around in the TODO. please add more words at least.